src/hotspot/share/gc/shenandoah/shenandoahForwarding.inline.hpp

```diff
@@ -55,7 +55,7 @@ inline oop ShenandoahForwarding::get_forwardee_mutator(oop obj) {
   shenandoah_assert_correct(NULL, obj);
   assert(Thread::current()->is_Java_thread(), "Must be a mutator thread");
 
-  markWord mark = obj->mark();
+  markWord mark = obj->mark_acquire();
```

Comment:
Actually, I would argue that you don't need acquire here, since you don't touch anything other than mark word, so that there is no order that needs to be enforced.

Reply:
It is not about the mark word. It is about transitive dependency: object copy stores -> forwardee installation (release) -> forwardee discovery (acquire) --> object copy access. Remember this: https://github.com/openjdk/jdk/pull/2498#discussion_r574498830

Reply:
Sorry. I commented on wrong place. I actually meant line #78 in try_update_forwardee().

Reply:
You have to do acquire there as well, because you can exit from the next block, the one with `old_mark.is_marked()`, never getting further. Then the whole thing unfolds: for transitive visibility of the object contents, you have to load the mark word with acquire before accessing forwardee contents.

Reply:
Yes. I wasn't thinking of the use of oop content after.